### ISSUE TITLE
nipypecli fails with command line node output pklz files

### ISSUE THREAD

**Issue by tashrifbillah** [nipy/nipype#3092]
### Summary
My pipeline connects a wrapper `Node` around `ls` and a standard `SelectFiles` `Node`. 
`nipypecli show from_Ls.pklz` shows error but 
`nipypecli show from_SelectFiles.pklz` is fine.

### Actual behavior

> nipypecli show /tmp/pklzError_pipeline/pklzError_pipeline/list_content/_node.pklz

```
191111-15:01:00,55 nipype.utils INFO:
         Running nipype version 1.2.3 (latest: 1.2.3)
191111-15:01:00,67 nipype.utils WARNING:
         No metadata was found in the pkl file. Make sure you are currently using the same Nipype version from the generated pkl.
Traceback (most recent call last):
  File "/home/tb571/miniconda3/bin/nipypecli", line 11, in <module>
    sys.exit(cli())
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/nipype/scripts/cli.py", line 108, in show
    pkl_data = loadpkl(pklz_file)
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/nipype/utils/filemanip.py", line 737, in loadpkl
    raise e
  File "/home/tb571/miniconda3/lib/python3.6/site-packages/nipype/utils/filemanip.py", line 720, in loadpkl
    unpkl = pickle.load(pkl_file)
AttributeError: Can't get attribute 'Ls' on <module '__main__' from '/home/tb571/miniconda3/bin/nipypecli'>

```

However, output of the following is fine:

> nipypecli show /tmp/pklzError_pipeline/pklzError_pipeline/select_files/_node.pklz

```
191111-15:01:58,301 nipype.utils INFO:
         Running nipype version 1.2.3 (latest: 1.2.3)
pklzError_pipeline.select_files
```

This behavior is also observed with `result_list_content.pklz` file. The `select_files` counterpart is fine though.

### Expected behavior
```
191111-15:01:58,301 nipype.utils INFO:
         Running nipype version 1.2.3 (latest: 1.2.3)
pklzError_pipeline.list_content
```

### How to replicate the behavior
I have curtailed my larger pipeline to the following simple script. Run that and see the output of above commands.

### Script/Workflow details

```
#!/usr/bin/env python

from nipype import Node, Workflow, SelectFiles
from nipype.interfaces.base import CommandLine, CommandLineInputSpec, TraitedSpec, Str

# ============================================================================================================
parent_directory= '/tmp/pklzError_pipeline/'


# ============================================================================================================
class LsInputSpec(CommandLineInputSpec):
    in_path= Str(argstr='%s')


class LsOutputSpec(TraitedSpec):
    pass

class Ls(CommandLine):
    _cmd = 'ls'
    input_spec = LsInputSpec
    output_spec = LsOutputSpec

    def _list_outputs(self):
        pass


# ============================================================================================================
templates= {'path': '/tmp/'}
select_files= Node(SelectFiles(templates), name= 'select_files')

# ============================================================================================================
pklzError_pipeline= Workflow(name='pklzError_pipeline', base_dir= parent_directory)

ls_node= Node(Ls(), name='list_content')

pklzError_pipeline.connect(select_files, 'path', ls_node, 'in_path')
pklzError_pipeline.run()

```

### Platform details:

```
191111-15:05:13,96 nipype.utils INFO:
         Running nipype version 1.2.3 (latest: 1.2.3)
{'commit_hash': 'a485cf60f',
 'commit_source': 'installation',
 'networkx_version': '2.2',
 'nibabel_version': '2.3.0',
 'nipype_version': '1.2.3',
 'numpy_version': '1.16.2',
 'pkg_path': '/home/tb571/miniconda3/lib/python3.6/site-packages/nipype',
 'scipy_version': '1.2.1',
 'sys_executable': '/home/tb571/miniconda3/bin/python',
 'sys_platform': 'linux',
 'sys_version': '3.6.7 | packaged by conda-forge | (default, Feb 28 2019, '
                '09:07:38) \n'
                '[GCC 7.3.0]',
 'traits_version': '4.6.0'}

```

**Comment by effigies** [comment]
Does this problem persist if you upgrade to 1.3.1?

**Comment by tashrifbillah** [comment]
Hmm, `pip install nipype` resolved to https://github.com/nipy/nipype/releases/tag/1.2.3 when I opened this issue.

It looks like there has been two more releases afterwards. I'll check your suggestion and let you know.

**Comment by tashrifbillah** [comment]
The problem persists. Tried with a fresh environment:

```
cd /tmp
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
./Miniconda3-latest-Linux-x86_64.sh -b -p `pwd`/miniconda3
source miniconda3/bin/activate
pip install nipype
./execute_the_above_reproducing_script.py

```

**Comment by satra** [comment]
i think this may come from lack of a `PYTHONPATH` there is no way for unpickle to figure out how the class is defined. 

can you try the following:

```
export PYTHONPATH=$(pwd)
./execute_the_above_reproducing_script.py
```

i'm assuming that `./execute_the_above_reproducing_script.py` contains the interface definition

in general python cannot unpickle things that are not somehow in its path (e.g., site-packages, PYTHONPATH).

**Comment by tashrifbillah** [comment]
Hi @satra , please see my edit [here](https://github.com/nipy/nipype/issues/3092#issuecomment-553560951).

Nonetheless, I tried your suggestion, but no fruit:

```
export PYTHONPATH=$(pwd)
./execute_the_above_reproducing_script.py
nipypecli show /tmp/pklzError_pipeline/pklzError_pipeline/select_files/_node.pklz
```

**Comment by satra** [comment]
can you change the script to the following:

<details>
<summary>New script</summary>

```
#!/usr/bin/env python

from nipype import Node, Workflow, SelectFiles
from nipype.interfaces.base import CommandLine, CommandLineInputSpec, TraitedSpec, Str

# ============================================================================================================
parent_directory= '/tmp/pklzError_pipeline/'


# ============================================================================================================
class LsInputSpec(CommandLineInputSpec):
    in_path= Str(argstr='%s')


class LsOutputSpec(TraitedSpec):
    pass

class Ls(CommandLine):
    _cmd = 'ls'
    input_spec = LsInputSpec
    output_spec = LsOutputSpec

    def _list_outputs(self):
        pass


if __name__ == "__main__":
    # ============================================================================================================
    templates= {'path': '/tmp/'}
    select_files= Node(SelectFiles(templates), name= 'select_files')

    # ============================================================================================================
    pklzError_pipeline= Workflow(name='pklzError_pipeline', base_dir= parent_directory)

    ls_node= Node(Ls(), name='list_content')

    pklzError_pipeline.connect(select_files, 'path', ls_node, 'in_path')
    pklzError_pipeline.run()
```

</details>

**Comment by tashrifbillah** [comment]
I should have mentioned, I dug into the traceback and tried a few apparent fixes, including the one you suggested. Didn't work.

**Comment by satra** [comment]
@tashrifbillah - i was able to run the script above that i pasted even without the PYTHONPATH on both miniconda on osx and the current nipype docker container: `nipype/nipype:py36`

<details><summary>OSX script output</summary>

```
$ python test.py
191113-15:01:26,273 nipype.utils INFO:
	 Running nipype version 1.4.0-dev+g2012092aa (latest: 1.3.1)
191113-15:01:27,605 nipype.workflow INFO:
	 Workflow pklzError_pipeline settings: ['check', 'execution', 'logging', 'monitoring']
191113-15:01:27,613 nipype.workflow INFO:
	 Running serially.
191113-15:01:27,613 nipype.workflow INFO:
	 [Node] Setting-up "pklzError_pipeline.select_files" in "/private/tmp/pklzError_pipeline/pklzError_pipeline/select_files".
191113-15:01:27,615 nipype.workflow INFO:
	 [Node] Running "select_files" ("nipype.interfaces.io.SelectFiles")
191113-15:01:27,634 nipype.workflow INFO:
	 [Node] Finished "pklzError_pipeline.select_files".
191113-15:01:27,635 nipype.workflow INFO:
	 [Node] Setting-up "pklzError_pipeline.list_content" in "/private/tmp/pklzError_pipeline/pklzError_pipeline/list_content".
191113-15:01:27,639 nipype.workflow INFO:
	 [Node] Running "list_content" ("__main__.Ls"), a CommandLine Interface with command:
ls /tmp/
191113-15:01:27,756 nipype.interface INFO:
	 stdout 2019-11-13T15:01:27.756049:5dcb94d641c12
191113-15:01:27,756 nipype.interface INFO:
	 stdout 2019-11-13T15:01:27.756049:adobegc.log
191113-15:01:27,756 nipype.interface INFO:
	 stdout 2019-11-13T15:01:27.756049:com.apple.launchd.U0cJi9HUcw
191113-15:01:27,756 nipype.interface INFO:
	 stdout 2019-11-13T15:01:27.756049:pklzError_pipeline
191113-15:01:27,756 nipype.interface INFO:
	 stdout 2019-11-13T15:01:27.756049:powerlog
191113-15:01:27,788 nipype.workflow INFO:
	 [Node] Finished "pklzError_pipeline.list_content".

$ python
Python 3.7.5 (default, Oct 25 2019, 10:52:18) 
```
</details>

<details><summary>Docker container script output</summary>

```
$ python test.py 
191113-20:07:48,296 nipype.utils INFO:
	 Running nipype version 1.4.0-dev (latest: 1.3.1)
191113-20:07:48,571 nipype.workflow INFO:
	 Workflow pklzError_pipeline settings: ['check', 'execution', 'logging', 'monitoring']
191113-20:07:48,574 nipype.workflow INFO:
	 Running serially.
191113-20:07:48,575 nipype.workflow INFO:
	 [Node] Setting-up "pklzError_pipeline.select_files" in "/tmp/pklzError_pipeline/pklzError_pipeline/select_files".
191113-20:07:48,577 nipype.workflow INFO:
	 [Node] Running "select_files" ("nipype.interfaces.io.SelectFiles")
191113-20:07:48,582 nipype.workflow INFO:
	 [Node] Finished "pklzError_pipeline.select_files".
191113-20:07:48,583 nipype.workflow INFO:
	 [Node] Setting-up "pklzError_pipeline.list_content" in "/tmp/pklzError_pipeline/pklzError_pipeline/list_content".
191113-20:07:48,587 nipype.workflow INFO:
	 [Node] Running "list_content" ("__main__.Ls"), a CommandLine Interface with command:
ls /tmp/
191113-20:07:48,604 nipype.interface INFO:
	 stdout 2019-11-13T20:07:48.604180:hsperfdata_root
191113-20:07:48,604 nipype.interface INFO:
	 stdout 2019-11-13T20:07:48.604180:pklzError_pipeline
191113-20:07:48,625 nipype.workflow INFO:
	 [Node] Finished "pklzError_pipeline.list_content".
neuro@65b5e2f49992:/data$ python
Python 3.6.7 | packaged by conda-forge | (default, Nov  6 2019, 16:19:42)
```
</details>

**Comment by satra** [comment]
hold on - i'm checking the cli

**Comment by satra** [comment]
this is because of pythonpath. the following works, it assumes that test.py is the script i posted earlier.

```
export PYTHONPATH=$(pwd)
rm -rf /tmp/pklzError_pipeline/pklzError_pipeline/
python test2.py # make sure to remove the old work
nipypecli show /tmp/pklzError_pipeline/pklzError_pipeline/list_content/_node.pklz 
```

<details><summary>test2.py</summary>

```
#!/usr/bin/env python

from nipype import Node, Workflow, SelectFiles
from nipype.interfaces.base import CommandLine, CommandLineInputSpec, TraitedSpec, Str

# ============================================================================================================
parent_directory= '/tmp/pklzError_pipeline/'
from test import Ls

if __name__ == "__main__":
    # ============================================================================================================
    templates= {'path': '/tmp/'}
    select_files= Node(SelectFiles(templates), name= 'select_files')

    # ============================================================================================================
    pklzError_pipeline= Workflow(name='pklzError_pipeline', base_dir= parent_directory)

    ls_node= Node(Ls(), name='list_content')

    pklzError_pipeline.connect(select_files, 'path', ls_node, 'in_path')
    pklzError_pipeline.run()

```
</details>

**Comment by tashrifbillah** [comment]
@satra, I removed old work before each of my testing.

> it assumes that test.py is the script i posted earlier

Can you rephrase?

Your output doesn't have anything from `nipypecli show`. I am having trouble with that, not running the pipeline.

> hold on - i'm checking the cli

I hope you tested the `nipypecli show` as well.

> miniconda on osx and the current nipype docker container

Do you think this is an RHEL Linux issue then? Here is my machine:

```
Linux 2 3.10.0-957.27.2.el7.x86_64 #1 SMP Mon Jul 29 17:46:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

Finally, I tested what your latest suggestion, didn't work.

**Comment by tashrifbillah** [comment]
You can try this container:

```
FROM centos:7.5.1804

MAINTAINER YouYou

# set up user and working directory
WORKDIR /
ENV USER="root"

copy test.py /

# libraries and commands for FSL
RUN yum -y update > /dev/null 2>&1 && \
    yum -y install epel-release wget which file bzip2 openblas-devel libquadmath && \
    yum clean all && \
# install python
    echo "Downloading miniconda3 installer" && \
    wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O Miniconda3-latest-Linux-x86_64.sh > /dev/null 2>&1 && \
    /bin/bash Miniconda3-latest-Linux-x86_64.sh -b -p miniconda3/ && \
    rm -f Miniconda3-latest-Linux-x86_64.sh && \
    source miniconda3/bin/activate && \
    echo "source $PWD/miniconda3/bin/activate" >> ~/.bashrc && \
# install nipype
    pip install nipype && \
    python /test.py && \
    nipypecli show /tmp/pklzError_pipeline/pklzError_pipeline/list_content/_node.pklz

```

**Comment by satra** [comment]
@tashrifbillah - you will need both `test.py` and `test2.py` (see above) and run with `python test2.py`

essentially you need a separate file (test.py) that has the interface, and import it into the script file (test2.py).

**Comment by tashrifbillah** [comment]
@satra , I did exactly what you said, yet I got the error. Did you get a chance to test with my container?

**Comment by satra** [comment]
@tashrifbillah - not quite the same

<details><summary>dockerfile</summary>

```
FROM centos:7.5.1804

MAINTAINER YouYou

# set up user and working directory
WORKDIR /
ENV USER="root"

copy test.py /
copy test2.py /

# libraries and commands for FSL
RUN yum -y update > /dev/null 2>&1 && \
    yum -y install epel-release gcc wget which file bzip2 openblas-devel libquadmath && \
    yum clean all && \
# install python
    echo "Downloading miniconda3 installer" && \
    wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O Miniconda3-latest-Linux-x86_64.sh > /dev/null 2>&1 && \
    /bin/bash Miniconda3-latest-Linux-x86_64.sh -b -p miniconda3/ && \
    rm -f Miniconda3-latest-Linux-x86_64.sh && \
    source miniconda3/bin/activate && \
    echo "source $PWD/miniconda3/bin/activate" >> ~/.bashrc && \
# install nipype
    pip install nipype
RUN source miniconda3/bin/activate && \
    export PYTHONPATH=$(pwd) && \
    python /test2.py && \
    nipypecli show /tmp/pklzError_pipeline/pklzError_pipeline/list_content/_node.pklz
```
</details>

<details><summary>output</summary>

```
$ docker build -t dbg .
Sending build context to Docker daemon  8.704kB
Step 1/8 : FROM centos:7.5.1804
 ---> cf49811e3cdb
Step 2/8 : MAINTAINER YouYou
 ---> Using cache
 ---> f3a0a4f1a979
Step 3/8 : WORKDIR /
 ---> Using cache
 ---> 9c9e0910b242
Step 4/8 : ENV USER="root"
 ---> Using cache
 ---> ba447d7ce026
Step 5/8 : copy test.py /
 ---> Using cache
 ---> dc3d79fc8a00
Step 6/8 : copy test2.py /
 ---> Using cache
 ---> 75dcd60787cd
Step 7/8 : RUN yum -y update > /dev/null 2>&1 &&     yum -y install epel-release gcc wget which file bzip2 openblas-devel libquadmath &&     yum clean all &&     echo "Downloading miniconda3 installer" &&     wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O Miniconda3-latest-Linux-x86_64.sh > /dev/null 2>&1 &&     /bin/bash Miniconda3-latest-Linux-x86_64.sh -b -p miniconda3/ &&     rm -f Miniconda3-latest-Linux-x86_64.sh &&     source miniconda3/bin/activate &&     echo "source $PWD/miniconda3/bin/activate" >> ~/.bashrc &&     pip install nipype
 ---> Using cache
 ---> dddab33dbfb3
Step 8/8 : RUN source miniconda3/bin/activate &&     export PYTHONPATH=$(pwd) &&     python /test2.py &&     nipypecli show /tmp/pklzError_pipeline/pklzError_pipeline/list_content/_node.pklz
 ---> Running in 4f4e8bea2754
191114-04:02:20,386 nipype.workflow INFO:
	 Workflow pklzError_pipeline settings: ['check', 'execution', 'logging', 'monitoring']
191114-04:02:20,389 nipype.workflow INFO:
	 Running serially.
191114-04:02:20,390 nipype.workflow INFO:
	 [Node] Setting-up "pklzError_pipeline.select_files" in "/tmp/pklzError_pipeline/pklzError_pipeline/select_files".
191114-04:02:20,392 nipype.workflow INFO:
	 [Node] Running "select_files" ("nipype.interfaces.io.SelectFiles")
191114-04:02:20,404 nipype.workflow INFO:
	 [Node] Finished "pklzError_pipeline.select_files".
191114-04:02:20,405 nipype.workflow INFO:
	 [Node] Setting-up "pklzError_pipeline.list_content" in "/tmp/pklzError_pipeline/pklzError_pipeline/list_content".
191114-04:02:20,407 nipype.workflow INFO:
	 [Node] Running "list_content" ("test.Ls"), a CommandLine Interface with command:
ls /tmp/
191114-04:02:20,423 nipype.interface INFO:
	 stdout 2019-11-14T04:02:20.423435:ks-script-3QMvMi
191114-04:02:20,423 nipype.interface INFO:
	 stdout 2019-11-14T04:02:20.423435:pklzError_pipeline
191114-04:02:20,423 nipype.interface INFO:
	 stdout 2019-11-14T04:02:20.423435:yum.log
191114-04:02:20,447 nipype.workflow INFO:
	 [Node] Finished "pklzError_pipeline.list_content".
pklzError_pipeline.list_content
Removing intermediate container 4f4e8bea2754
 ---> 3b832270d1ac
Successfully built 3b832270d1ac

```
</details>

**Comment by tashrifbillah** [comment]
Okay, sorry about that. Can you please post `test.py` again?

**Comment by satra** [comment]
test.py: https://github.com/nipy/nipype/issues/3092#issuecomment-553569582
test2.py: https://github.com/nipy/nipype/issues/3092#issuecomment-553583551

**Comment by tashrifbillah** [comment]
I realized during our conversation, I was defining PYTHONPATH only during `nipypecli show`, not before execution of `test2.py`. My apologies for bugging you this far. On the other hand, test file needs to contain only the definition of the interface, not the `if __name__=='__main__':` parts.
```
#!/usr/bin/env python

from nipype import Node, Workflow, SelectFiles
from nipype.interfaces.base import CommandLine, CommandLineInputSpec, TraitedSpec, Str

# ============================================================================================================
parent_directory= '/tmp/pklzError_pipeline/'


# ============================================================================================================
class LsInputSpec(CommandLineInputSpec):
    in_path= Str(argstr='%s')


class LsOutputSpec(TraitedSpec):
    pass

class Ls(CommandLine):
    _cmd = 'ls'
    input_spec = LsInputSpec
    output_spec = LsOutputSpec

    def _list_outputs(self):
        pass

```

However, it has come down to writing a separate file containing interface for each of the python scripts I am connecting with `workflow( )`. So far, I was putting all the interfaces in [one file](https://github.com/pnlbwh/pnlNipype/blob/nipype-workflow/scripts/eddy_cli_workflow.py). [These](https://github.com/pnlbwh/pnlNipype/tree/master/scripts) are our pipeline scripts, which are command line executables I am connecting with `nipype`. Given that, I am feeling a sense of cumbersomeness for connecting python executable scripts.

Feel free to close the issue.

If I am not following the best way to connect python scripts, please suggest as well.

**Comment by tashrifbillah** [comment]
Thanks for patiently supporting this issue.

**Comment by satra** [comment]
> On the other hand, test file needs to contain only the definition of the interface, not the if __name__=='__main__': parts.

that is correct. the if part doesn't get executed when the file is loaded as a module.

the issues you are running into are basic python constructs, nothing to do with nipype per se. you would have the same kinds of things if you tried to pickle python functions or classes defined in a script. the interpreter needs to know where (the actual path of the module) the function/class comes from. 

there is a python package called cloudpickle which gets around this, and we will likely move to that for a future version of nipype. but at present you will have to separate interfaces in a different file, and import it into your script.

**Comment by tashrifbillah** [comment]
@satra , an FYI for you: I didn't have to write individual definition file for each of my python scripts I am attempting to glue together by `nipype`. Instead, I have been able to put my definitions in just one file. So, the only additional work than before is, other than having the definitions in the file where they are used, I had to put all of them in a separate file. I stand corrected, it is not cumbersome. :)
